src/PortingAssistant.Client.Telemetry/Uploader.cs

```diff
@@ -12,130 +10,251 @@
+                        continue;
                     }
-                    else if (fileExtension == ".metrics")
+
+                    File.Delete(file.FullName);
+                    if (_fileLineNumberMap.ContainsKey(file.FullName))
                     {
-                        logName = $"{logName}-metrics";
+                        updateLastTokenJson = true;
+                        _fileLineNumberMap.Remove(file.FullName);
                     }
+                }
+
+                if (updateLastTokenJson)
+                {
+                    UpdateFileLineMapJson();
+                }
+            }
+            catch (Exception e)
+            {
+                ErrorList.Add("CleanupLogFolderError", e);
+            }
+        }
+
+        private void ReadFileLineMap()
+        {
+            _lastReadTokenFile = Path.Combine(_configuration.LogsPath, "lastToken.json");
+            if (File.Exists(_lastReadTokenFile))
+            {
+                _fileLineNumberMap = JsonConvert.DeserializeObject<Dictionary<string, int>>(File.ReadAllText(_lastReadTokenFile));
+                if (_fileLineNumberMap == null)
+                {
+                    _fileLineNumberMap = new Dictionary<string, int>();
+                }
+            }
+        }
+
+        private void UpdateFileLineMapJson()
+        {
+            ReadFileLineMap();
+            foreach (var pair in _updatedFileLineNumberMap)
+            {
+                if (_fileLineNumberMap.ContainsKey(pair.Key))
+                {
+                    _fileLineNumberMap[pair.Key] = pair.Value;
+                }
+                else
+                {
+                    _fileLineNumberMap.Add(pair.Key, pair.Value);
+                }
+            }
+            using FileStream fs = WaitForFile(_lastReadTokenFile,
+                FileMode.OpenOrCreate,
+                FileAccess.ReadWrite,
+                FileShare.Read);
+            using StreamWriter writer = new StreamWriter(fs);
+            writer.Write(JsonConvert.SerializeObject(_fileLineNumberMap));
+        }
 
-                    // Add new files to fileLineNumberMap
-                    if (!fileLineNumberMap.ContainsKey(file))
+        private FileStream WaitForFile(string fullPath,
+            FileMode mode,
+            FileAccess access,
+            FileShare share)
+        {
+            for (int numTries = 0; numTries < 10; numTries++)
+            {
+                FileStream fs = null;
+                try
+                {
+                    fs = new FileStream(fullPath, mode, access, share);
+                    return fs;
+                }
+                catch (IOException)
+                {
+                    if (fs != null)
                     {
-                        fileLineNumberMap[file] = 0;
+                        fs.Dispose();
                     }
-                    var initLineNumber = fileLineNumberMap[file];
+                    Thread.Sleep(50);
```

Comment:
We are waiting for 50ms only - which might not have an impact. I suggest less retries over a longer time.

Reply:
updated to 5 second wait over 5 tries